### PR TITLE
LC-14 - agency-self-sign-up

### PR DESCRIPTION
**Pull request by m24murray**
- domain, controller, repository for agency tokens

### CODE REVIEW

File: src/main/java/uk/gov/cshr/civilservant/service/OrganisationalUnitService.java [hunk 1]
```diff
@@ -43,13 +46,41 @@ private void getParent(OrganisationalUnit organisationalUnit, List<Organisationa
...
+        }
+
+        currentToken.setAgencyDomains(newToken.getAgencyDomains());
+        currentToken.setCapacity(newToken.getCapacity());
+        currentToken.setToken(newToken.getToken());
+
+        return repository.save(organisationalUnit);
+    }
+
+    public OrganisationalUnit deleteAgencyToken(OrganisationalUnit organisationalUnit) {
```
Comment (Daniel-Keenan) [inline src/main/java/uk/gov/cshr/civilservant/service/OrganisationalUnitService.java]: Will need to add code to delete token from the database as part of this method

**Comment by m24murray** [comment]
can we get rid of the `.DS_Store` file?

File: src/main/java/uk/gov/cshr/civilservant/controller/AgencyTokenController.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    " code=" + updateSpacesForAgencyTokenRequestDTO.getCode() +
+                    " isRemoveUser=" + updateSpacesForAgencyTokenRequestDTO.isRemoveUser());
+
+            List<OrganisationalUnit> passedOrganisationalUnitList = new ArrayList<>();
+            List<String> organisationalUnitCodeList = new ArrayList<>();
+            passedOrganisationalUnitList = organisationalUnitService.getOrganisationWithParents(updateSpacesForAgencyTokenRequestDTO.getCode());
+            for (OrganisationalUnit organisationalUnit : passedOrganisationalUnitList) {
+                organisationalUnitCodeList.add(organisationalUnit.getCode());
+            }
+            agencyTokenService.updateAgencyTokenSpacesAvailable(updateSpacesForAgencyTokenRequestDTO.getDomain(), updateSpacesForAgencyTokenRequestDTO.getToken(), organisationalUnitCodeList, updateSpacesForAgencyTokenRequestDTO.isRemoveUser());
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/AgencyTokenController.java]: Could we replace lines 69-74 with something like the following:

```
List<String> organisationalUnitCodeList = organisationalUnitService.getOrganisationWithParents(updateSpacesForAgencyTokenRequestDTO.getCode())
                    .stream()
                    .map(OrganisationalUnit::getCode)
                    .collect(Collectors.toList());
```

Should tidy up and reduce number of vars being created

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java [hunk 3]
```diff
@@ -105,6 +114,83 @@ public CivilServantController(LineManagerService lineManagerService, CivilServan
         return ResponseEntity.unprocessableEntity().build();
     }
 
+    @GetMapping("/org")
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: These organisation methods should maybe be moved to another class. It make sense for `CivilServantController` to only contain methods pertaining to the CivilServant domain objects.

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java [hunk 4]
```diff
@@ -39,18 +44,22 @@
...
     public CivilServantController(LineManagerService lineManagerService, CivilServantRepository civilServantRepository,
                                   RepositoryEntityLinks repositoryEntityLinks,
-                                  CivilServantResourceFactory civilServantResourceFactory) {
+                                  CivilServantResourceFactory civilServantResourceFactory,
+                                  OrganisationalUnitRepository organisationalUnitRepository) {
         this.lineManagerService = lineManagerService;
         this.civilServantRepository = civilServantRepository;
         this.repositoryEntityLinks = repositoryEntityLinks;
         this.civilServantResourceFactory = civilServantResourceFactory;
+        this.organisationalUnitRepository = organisationalUnitRepository;
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: See comment below, i dont think we should be calling the organisationalUnitRepository directly from the CivilServantController.

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java [hunk 5]
```diff
@@ -105,6 +114,83 @@ public CivilServantController(LineManagerService lineManagerService, CivilServan
...
+
+    @PatchMapping("/org")
+    @PreAuthorize("isAuthenticated()")
+    public ResponseEntity updateOrganisation(@Valid @RequestBody UpdateOrganisationDTO updateOrganisationDTO) {
+
+        log.info("updating civil servants organisation for organisation=" + updateOrganisationDTO.getOrganisation());
+
+        try {
+            Optional<CivilServant> optionalCivilServant = civilServantRepository.findByPrincipal();
+            if (optionalCivilServant.isPresent()) {
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: Replacing your `Optional.isPresent()` and `Optional.get()` with one of the following allows you a more flexible method of handling conditional logic. 
```
Optional.isPresent(obj -> {})
Optional.map(obj -> {})
Optional.orElseGet(obj -> {})
```

File: src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java [hunk 6]
```diff
@@ -40,12 +51,36 @@ public OrganisationalUnitController(OrganisationalUnitService organisationalUnit
...
     }
 
+    @GetMapping("/flat/{domain}/")
+    public ResponseEntity<List<OrganisationalUnitDto>> listOrganisationalUnitsAsFlatStructureFilteredByDomainAndCode(@PathVariable String domain) {
+
+        log.info("Getting org flat, filtered by domain");
+        try {
+            List<OrganisationalUnit> organisationalUnits = organisationalUnitService.getOrganisationsForDomain(domain);
+            if (!organisationalUnits.isEmpty()) {
+                List<OrganisationalUnitDto> dtos = organisationalUnits.stream().map(ou -> organisationalUnitDtoFactory.create(ou)).collect(Collectors.toList());
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java]: If you're doing a stream/map on a list, you might not need the `isEmpty` check on line 66

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java [hunk 7]
```diff
@@ -105,6 +114,83 @@ public CivilServantController(LineManagerService lineManagerService, CivilServan
...
+        log.debug("Getting civil servant org details for user with uid " + uid);
+
+        Optional<CivilServant> civilServant = civilServantRepository.findByIdentity(uid);
+
+        if (civilServant.isPresent()) {
+            return civilServantResourceFactory.getCivilServantOrganisationalUnitCode(civilServant.get())
+                    .map(orgCodeDTO -> ResponseEntity.ok(orgCodeDTO))
+                    .orElse(buildNotFound(uid));
+        } else {
+            return buildNotFound(uid);
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: I'd prefer not to refactor the return of the `ResponseEntity` into a method, as it should be clear what the return of the controller method is.

File: src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java [hunk 8]
```diff
@@ -72,5 +107,78 @@ public OrganisationalUnitController(OrganisationalUnitService organisationalUnit
...
+        ex.getBindingResult().getAllErrors().forEach((error) -> {
+            String fieldName = ((FieldError) error).getField();
+            String errorMessage = error.getDefaultMessage();
+            errors.put(fieldName, errorMessage);
+        });
+        return errors;
+    }
+
+    /* assumes any validation has already happened.*/
+    private AgencyToken buildAgencyTokenFromAgencyTokenDTO(AgencyTokenDTO agencyTokenDTO, boolean isCreateNewToken) {
```
Comment (m24murray) [inline src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java]: Can extract into a factory class?

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -48,13 +48,20 @@ oauth:
...
 govNotify:
   enabled: ${GOV_NOTIFY_ENABLED:false}
   key: ${GOV_NOTIFY_API_KEY:changeme}
   template:
     lineManager: ${GOV_NOTIFY_LINEMANAGER_TEMPLATE_ID:changeme}
 
+#Validation properties
+agencyToken:
+  capacity:
+    max: 1500
```
Comment (m24murray) [inline src/main/resources/application.yml]: These will need to be configurable at some point

File: src/main/resources/db/migration/h2/V1.13__add-force-org-reset-flag.sql~HEAD [hunk 10]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `civil_servant` ADD `force_org_reset` bit(1) DEFAULT FALSE;
```
Comment (m24murray) [inline src/main/resources/db/migration/h2/V1.13__add-force-org-reset-flag.sql~HEAD]: Think this is a redundant file from the rebasing

File: src/main/java/uk/gov/cshr/civilservant/controller/AgencyTokenController.java (hunk 2 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/AgencyTokenController.java]: Updated.

**Comment by brdgeth** [comment]
> can we get rid of the `.DS_Store` file?

I don't have this myself.  However I have added it to the .gitignore file.
If anyone still has it locally they need to remove it from the git cache.

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java (hunk 3 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: As discussed, this is parked until we have more time.

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java (hunk 4 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: See above

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java (hunk 5 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: Updated.

File: src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java (hunk 7 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/CivilServantController.java]: Updated

File: src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java (hunk 6 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java]: Updated

File: src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java (hunk 8 above)
Comment (brdgeth) [reply src/main/java/uk/gov/cshr/civilservant/controller/OrganisationalUnitController.java]: Updated

File: src/main/resources/db/migration/h2/V1.13__add-force-org-reset-flag.sql~HEAD (hunk 10 above)
Comment (brdgeth) [reply src/main/resources/db/migration/h2/V1.13__add-force-org-reset-flag.sql~HEAD]: I spotted this when I was checking the code the other week ago.  Already deleted the duplicate file.

File: src/main/resources/application.yml (hunk 9 above)
Comment (brdgeth) [reply src/main/resources/application.yml]: The application.yml does mean they are configurable, although it depends on the DevOps.  I am not familiar with the DevOps on this project.  Therefore please can you explain what you mean by configurable.

Comment (m24murray) [reply src/main/resources/application.yml]: Agree its a bit pointless, but to be configurable via env variable has been requested in requirements: 
e.g. ${AGENCY_TOKEN_MAX_CAPACITY:1500}

Comment (brdgeth) [reply src/main/resources/application.yml]: ah ok, i see what you mean.  Thank you.  Ill add it.

Comment (brdgeth) [reply src/main/resources/application.yml]: Updated

**Comment by brdgeth** [comment]
I have updated all the items you mentioned.  Except the update method which has been parked for now.

**Pull request merged**